### PR TITLE
rfm22b: Allow selection of hardware band.

### LINKED ISSUE #1761: RFM22B frequency band should be settable
There were incomplete attempts at this in #1610 and #1655 .

Now with the hardware refactor pending in #1759 it will be easier to do this, but it will require redoing some of this code.

Closing the underlying pull requests and opening this issue for tracking.

### PR DESCRIPTION
**Pull request by mlyle**
For both Tau-style telemetry and OpenLRS.  Initial frequencies chosen
for Tau-style telemetry on 868 and 915 bands.

Fixes #1761

### CODE REVIEW

**Comment by mlyle** [comment]
Needs test and verification help.  I will cover 433MHz cases.

**Comment by peabody124** [comment]
Can we merge #1791 first since that was blocked on the HAL and is now rebased.

**Comment by mlyle** [comment]
Sure.  This is blocked on test, and I don't expect significant conflict from #1791.

**Comment by peabody124** [comment]
Yeah now I saw it's just a one-liner conflict I'm not bothered so it can merge in whichever order they are validated :)

**Comment by mlyle** [comment]
Rebased to next because of conflict from #1791.

**Comment by mlyle** [comment]
@sfabris @readerror67 @DTFUHF 

If this is going to merge, it needs test very soon.

**Comment by sfabris** [comment]
@myle Unfortunatelly I don't think I'll be able to test this in the next weeks. Work, wife and kids are not leaving enough free time :-)

**Comment by mlyle** [comment]
Rebased to remove conflicts on next.

File: shared/uavobjectdefinition/hwshared.xml [hunk 1]
```diff
@@ -34,6 +34,7 @@
 		<field name="MaxRfSpeed" units="bps" type="enum" elements="1" options="9600,19200,32000,64000,100000,192000" defaultvalue="64000"/>
 		<field name="MaxRfPower" units="mW" type="enum" elements="1" options="0,1.25,1.6,3.16,6.3,12.6,25,50,100" defaultvalue="1.25"/>
 		<field name="DSMxMode" units="mode" type="enum" elements="1" options="Autodetect,Force 10-bit,Force 11-bit,Bind 3 pulses,Bind 4 pulses,Bind 5 pulses,Bind 6 pulses,Bind 7 pulses,Bind 8 pulses,Bind 9 pulses,Bind 10 pulses" defaultvalue="Autodetect"/>
+		<field name="RfBand" units="MHz" type="enum" elements="1" options="BoardDefault,433,868,915" defaultvalue="BoardDefault"/>
```
Comment (peabody124) [inline shared/uavobjectdefinition/hwshared.xml]: the board default value doesn't seem to be handled properly. it actually seems to fall through directly to 433 https://github.com/TauLabs/TauLabs/pull/1803/files#diff-2463d1260e83f0034184987f54d23dc5R109 so it's not really a board default as much as a default to 433. Could we remove this value entirely and then in the hwsparky2.xml for example the default value there defines the board default?

File: flight/PiOS/Common/pios_openlrs.c [hunk 2]
```diff
@@ -104,6 +104,54 @@ const uint8_t default_hop_list[] = {DEFAULT_HOPLIST};
 const struct rfm22_modem_regs bind_params =
 { 9600, 0x05, 0x40, 0x0a, 0xa1, 0x20, 0x4e, 0xa5, 0x00, 0x20, 0x24, 0x4e, 0xa5, 0x2c, 0x23, 0x30 };
 
+static uint32_t minFreq(HwSharedRfBandOptions band) {
+	switch (band) {
+		default:
```
Comment (peabody124) [inline flight/PiOS/Common/pios_openlrs.c]: and if you remove the board default value below you could just trash the default here so we get a compile failure if not covering any cases in the future.

File: flight/PiOS/Common/pios_rfm22b.c [hunk 3]
```diff
@@ -407,6 +411,25 @@ int32_t PIOS_RFM22B_Init(uint32_t * rfm22b_id, uint32_t spi_id,
 	rfm22b_dev->slave_num = slave_num;
 	rfm22b_dev->spi_id = spi_id;
 
+	uint32_t base_freq;
+
+	// And the frequency
+	switch (band) {
+		case HWSHARED_RFBAND_BOARDDEFAULT:
```
Comment (peabody124) [inline flight/PiOS/Common/pios_rfm22b.c]: again i'd be inclined to dropping this line

**Comment by peabody124** [comment]
Code changes look good. I don't have any non-433 hardware to test but can test this with OLRS and TLM

**Comment by peabody124** [comment]
Flight tested with both OLRS and TLM on BrushedSparky, including the binding process. Good to merge for me, although I'd prefer to see the BoardDefault dropped since I don't think it means what it says or adds functionality.

**Comment by mlyle** [comment]
I'd like to see this work on another band before merge.

**Comment by peabody124** [comment]
I'm happy to merge this without testing given how long it's sat here and the code review. What are your thoughts?

**Pull request closed without merging**